### ISSUE TITLE
How to fix 2:4 mask after loading weights on SparseLinear based on this repo

### ISSUE THREAD

**Issue by erwangccc** [NM-sparsity/NM-sparsity#16]
Hi, @sunlex0717 @unique-null @NM-sparsity 

Sorry to bother you.

I want to fix mask after i load pre-trained weights based on 2:4 sparsity, and the location of mask will not change during training, like nvidia's approach. And how to implement it based on this repo. 

I have tried ASP from apex. But there are too much errors when using ASP at swin-transformer model. Maybe NV‘ ASP does not support transformers well.

Thanks in advance~

**Comment by aojunzz** [comment]
hi @erwangccc 

Based on this repo, you can apply the mask on gradients, and the mask will not change the location.

Aojun

**Comment by erwangccc** [comment]
Hi, @NM-sparsity 

Thanks for your reply.

As the following code shows, i think the mask will be updated during training, it will change sparsity location, right? SR-STE will do sparse while training.
```
class Sparse(autograd.Function):
    """" Prune the unimprotant weight for the forwards phase but pass the gradient to dense weight using SR-STE in the backwards phase"""

    @staticmethod
    def forward(ctx, weight, N, M, decay = 0.0002):
        ctx.save_for_backward(weight)

        output = weight.clone()
        length = weight.numel()
        group = int(length/M)

        weight_temp = weight.detach().abs().reshape(group, M)
        index = torch.argsort(weight_temp, dim=1)[:, :int(M-N)]

        w_b = torch.ones(weight_temp.shape, device=weight_temp.device)
        w_b = w_b.scatter_(dim=1, index=index, value=0).reshape(weight.shape)
        ctx.mask = w_b
        ctx.decay = decay

        return output*w_b
```

So i wander that how to solidify mask 'w_b' when i load weights first time then just the weights according to this location do backward.

**Comment by aojunzz** [comment]
what about your trained model?

**Comment by erwangccc** [comment]
Hi, @NM-sparsity 
It's Swin-T. I want to change a few ops such as conv2d or linear and train the model. 
I'm finetuning this backbone following your approach, modify some linear ops. The result shows that it can reach original acc well. So i want to do some training after solidifying mask as NV's approach so that to compare with last result.

Thanks～

**Comment by aojunzz** [comment]
@erwangccc hi you can add a self.mask in __init__ function of sparseConv/SparseLinear layers, and the self.mask initialize from the self.weight pruning. And then you apply the self.mask on self.weight (w = self.weight * self.mask). You don't need use " w = self.get_sparse_weights()"

**Comment by erwangccc** [comment]
Hi, @NM-sparsity 

Thanks, I have implemented the approach like you mentioned above. 
By the way,  the code as following, why do you use `(1 - ctx.mask) * weight`  to do backward? I think it will backpropagate the smaller N weights. it should be `ctx.mask * weight`. We should backpropagate higher weights, right?
```
    @staticmethod
    def backward(ctx, grad_outputs):
        weight, = ctx.saved_tensors
        return grad_outputs + ctx.decay * (1 - ctx.mask) * weight, None, None, None
```

**Comment by unique-null** [comment]
> Hi, @NM-sparsity
> 
> Thanks, I have implemented the approach like you mentioned above. By the way, the code as following, why do you use `(1 - ctx.mask) * weight` to do backward? I think it will backpropagate the smaller N weights. it should be `ctx.mask * weight`. We should backpropagate higher weights, right?
> 
> ```
>     @staticmethod
>     def backward(ctx, grad_outputs):
>         weight, = ctx.saved_tensors
>         return grad_outputs + ctx.decay * (1 - ctx.mask) * weight, None, None, None
> ```

Try to update the mask index dynamically, avoiding the frozen mask index @erwangccc

**Comment by erwangccc** [comment]
Hi, @unique-null @NM-sparsity 

I have understand why you add the weight from zero position, but another question, why you add the weight directly, not gradient.

Thank you in advance.
